Watchman.Discord/Areas/Users/Controllers/WarnsController.cs

```diff
@@ -0,0 +1,100 @@
+{
+    public class WarnsController : IController
+    {
+        private readonly MessagesServiceFactory _messagesServiceFactory;
+        private readonly UsersService _usersService;
+        private readonly DirectMessagesService _directMessagesService;
+        private readonly WarnService _warnService;
+
+        public WarnsController(MessagesServiceFactory messagesServiceFactory, UsersService usersService, DirectMessagesService directMessagesService, WarnService warnService)
+        {
+            this._messagesServiceFactory = messagesServiceFactory;
+            this._usersService = usersService;
+            this._directMessagesService = directMessagesService;
+            this._warnService = warnService;
+        }
+
+        [AdminCommand]
+        public async Task AddWarn(AddWarnCommand command, Contexts contexts)
+        {
+            var msgService = _messagesServiceFactory.Create(contexts);
+
+            if (contexts.User.IsAdmin)
+            {
+                if (string.IsNullOrEmpty(command.Reason) || string.IsNullOrWhiteSpace(command.Reason))
+                {
+                    await msgService.SendResponse(x => x.EmptyArgument("-reason"));
+                } 
+                else
+                {
+                    var mentionedUser = _usersService.GetUserByMention(contexts.Server, command.User);
+                    await _warnService.AddWarnToUser(command, contexts, mentionedUser);
+                    await msgService.SendResponse(x => x.UserHasBeenWarned(contexts.User.Name, mentionedUser.Name, command.Reason));
+                }
+            }
+            else
+            {
+                await msgService.SendResponse(x => x.NotAdminPermissions());
+            }
+        }
+
+        public async Task Warns(WarnsCommand command, Contexts contexts)
+        {
+            var mentionedUser = command.User == null ? contexts.User : _usersService.GetUserByMention(contexts.Server, command.User);
+            var serverId = command.All ? 0 : contexts.Server.Id;
+            var msgService = _messagesServiceFactory.Create(contexts);
+
+            if (command.All)
+            {
+                if (contexts.User.IsAdmin)
+                {
+                    if (mentionedUser == null)
+                    {
+                        await _directMessagesService.TrySendMessage(contexts.User.Id, x => x.UserNotFound(command.User), contexts);
+                    }
+                    else
+                    {
+                        var warnsStr = await _warnService.GetWarnsToString(serverId, mentionedUser.Id);
+                        await _directMessagesService.TrySendMessage(contexts.User.Id, x => x.GetUserWarns(mentionedUser.Name, warnsStr), contexts);
+                    }
+                }
+                else
+                {
+                    throw new NotAdminPermissionsException();
+                }
+            }
+            else
```

Comment:
taka dobra praktyka zwiększająca czytelność to unikanie else
jeśli kod jest bardzo mały to nie ma sensu się do tego stosować
(jeśli zarówno if i else są bardzo małe, np 2-3 linijkowe)

to męczące jeśli masz potężnego ifa i jeszcze pod nim else
i musisz sie zastanawiać jeszcze raz co robił ten if, żeby zrozumieć kiedy odpala sie else 
szczególnie jeśli masz ifa w ifie, else w else itd 
tworzy się wtedy takie drzewko
z logiką to ogarnięcia w 5 sekund
ale czytelność kodu wymaga zastanawiania sie przez duużo dłużej

Reply:
ok chyba polepszyłem trochę